### DISCUSSION TITLE
XML and End of line management

### DISCUSSION THREAD

**Discussion by caiapha** [community/community#77438] [New to GitHub] [answered]
Hello to all! 
I'm Caiapha and I'm totally noob on GitHub. I'm trying to create additional contents for a software which uses XML files as input. Unfortunately I'm totally noob also on XML. I just know some HTML and C++. 

My work is mostly to "copy&past" from existing files, adapting them to my needings. GitHub contains the reference input files, which are downloaded by the software in case of uptades.

My probem is related to End of Line character: the downloaded files uses UNIX EOL character, while I need Windows (CRLF) EOL.
I understood it should be possible to change the setting by a .gitattributes file, but all my tentatives failed.

the standard xml code to manage is like the following:
<sheet>
    <description> text 
                            text
                            text
     </description>
</sheet>

which is printed this way: [text text text]

Can someone help me to solve this problem?

thanks in advance!

**Comment by b8zeek** [comment]
Greetings Caiapha, it's nice to have you here buddy. 👋🏼 

You can use a `.gitattributes` file to manage line endings in your repository. Follow the steps below:
- create a new file in the root `.gitattributes` in case it doesn't exist
- add the following lines to the file:
```
* text=auto
*.xml text eol=crlf
```
- commit and push the `.gitattributes` to your repository.

This will tell Git to automatically manage line endings for all text files, and to use CRLF line endings for XML files, when you checkout an XML file, Git converts the line endings to CRLF, when you commit the file, Git converts the line endings back to LF.

One more thing, if you already have XML files in your repository, you might need to re-normalize them by running the commands:

```
git add --renormalize .
git commit -m "Normalized line endings"
```

Let me know if this solved the issues; best of luck on the journey! 🍺 

🍀

**Comment by caiapha** [comment]
dear Bejzik8, thanks for the fast reply.
I tried to follow your suggetion, but it didn't work for me. To be precise, I made the following:

1 I deleted all the files from my repository
2 I created a new file, named .gitattributes with inside the two line code you suggested
3 I created new files, in which I copy&paste the content of the files from my local folder. I used the "paste text only", not the generic "paste"
4 I deleted all the local files
5 I lauched the update, downloaded the files and restarted the software (as automatically required by itself)

 Unfortunately it failed.

Regarding the "renormalize", I have no Idea how to run that codes. Is there any shell, command line? 
tnx

**Reply by b8zeek** [reply]
Yeah, you run it from the command line.

🍀

**Reply by caiapha** [reply]
Where is it? I cannot find any command line. I'm working on browser.

**Reply by caiapha** [reply, marked as answer]
SOLVED! The problem was **probably** related to the GIT's interpretation of the "text" areas. In practice, what I need to consider consider as text was not interpreted as text, therefore the eol was not changed. 
I deleted the "text" word from "*.xml text eol=crlf" command line in .gitattribute and the file has been setted as windows style, solving the problem (for me).
I don't know if this is a "deprecated" solution, but it works to me.
Anyway, thanks for the support!

**Reply by queenofcorgis** [reply]
@caiapha excellent work thanks for coming back to share the solve incredible 👏🏼

**Comment by fahmizainal17** [comment]
Could you guys help me get two answer accepted?

**Reply by b8zeek** [reply]
Can you answer the questions and help the people?

🍀

**Reply by queenofcorgis** [reply]
@fahmizainal17 GitHub's [Acceptable Use Policies](https://docs.github.com/en/site-policy/acceptable-use-policies/github-acceptable-use-policies) prohibits coordinated or inauthentic activity like rapid questions and answers.  Please note any violations may result in a temporary or indefinite block from the Community. Thanks for understanding.
